### PR TITLE
Feature/encrypted dump

### PR DESCRIPTION
**Pull request by fceller**

### CODE REVIEW

File: arangosh/Dump/DumpFeature.cpp [hunk 1]
```diff
@@ -513,16 +523,21 @@ int DumpFeature::runDump(std::string& dbName, std::string& errorMsg) {
...
+      endEncryption(fd);
       TRI_TRACKED_CLOSE_FILE(fd);
       errorMsg = "cannot write to file '" + fileName + "'";
 
       return TRI_ERROR_CANNOT_WRITE_FILE;
     }
 
+    endEncryption(fd);
     TRI_TRACKED_CLOSE_FILE(fd);
   } catch (...) {
```
Comment (jsteemann) [inline arangosh/Dump/DumpFeature.cpp]: minor improvement:
the statements:
```
  endEncryption(fd);
  TRI_TRACKED_CLOSE_FILE(fd);
```
will be executed inside the `if` condition but also after the if:
```
if (!result) {
  endEncryption(fd);
  TRI_TRACKED_CLOSE_FILE(fd);
  ...
  return ...
}
endEncryption(fd);
TRI_TRACKED_CLOSE_FILE(fd);
```
we should move them before the if then:
```
endEncryption(fd);
TRI_TRACKED_CLOSE_FILE(fd);

if (!result) {
  ...
  return ...
}
```

File: arangosh/Dump/DumpFeature.cpp [hunk 2]
```diff
@@ -615,16 +630,21 @@ int DumpFeature::runDump(std::string& dbName, std::string& errorMsg) {
...
+        endEncryption(fd);
         TRI_TRACKED_CLOSE_FILE(fd);
         errorMsg = "cannot write to file '" + fileName + "'";
 
         return TRI_ERROR_CANNOT_WRITE_FILE;
       }
 
+      endEncryption(fd);
       TRI_TRACKED_CLOSE_FILE(fd);
     }
```
Comment (jsteemann) [inline arangosh/Dump/DumpFeature.cpp]: same here

File: arangosh/Dump/DumpFeature.cpp [hunk 3]
```diff
@@ -921,16 +945,21 @@ int DumpFeature::runClusterDump(std::string& errorMsg) {
...
```
Comment (jsteemann) [inline arangosh/Dump/DumpFeature.cpp]: same here

**Comment by neunhoef** [comment]
One more thing: If one tries to restore an encrypted dump without the key, a cryptic error message appears rather than saying "this is an encrypted dump, please use --encryption.keyfile" to specify the secret key. Similarly, a try to restore an unencrypted dump with an encryption key should gracefully fail. After all, we have written the ENCRYPTION file to recognize the situation.

**Comment by neunhoef** [comment]
Other than that, it seems to work.

**Pull request merged**